### ISSUE TITLE
Substitutions result in unbacked SymInts showing up before their definition sites

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#123854]
### 🐛 Describe the bug

Internal xref: https://fb.workplace.com/groups/6829516587176185/posts/7046138665513975/

This fails WITHOUT https://github.com/pytorch/pytorch/pull/120816

I don't entirely understand how to reproduce this problem, but in the live traces the invariant violation is fairly clear, so let's try to reason it out this way.

The proximal error when running the full reproducer quoted above is:

```
  File "/data/users/ezyang/fbsource/buck-out/v2/gen/fbcode/410ce23094e3dfef/torchrec/distributed/tests/__test_pt2_train__/test_pt2_train#link-tree/torch/_
inductor/scheduler.py", line 1302, in __init__                                                                                                            
    self.compute_dependencies()                                                                                                                           
  File "/data/users/ezyang/fbsource/buck-out/v2/gen/fbcode/410ce23094e3dfef/torchrec/distributed/tests/__test_pt2_train__/test_pt2_train#link-tree/torch/_
inductor/scheduler.py", line 1542, in compute_dependencies                                                                                                
    s in unbacked_symbol_to_origin_node                                                                                                                   
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:                                                                                       
AssertionError: u42 not in { ...
```

This assert failure means that we have a node that depends on u42, but at the time we are trying to schedule it, we haven't seen any node (e.g., an item call) that actually defines u42.

Inspection of the graph reveals that a renaming disaster has occurred:

```
        _local_scalar_dense_38: "Sym(u356)" = torch.ops.aten._local_scalar_dense.default(select_37);  select_37 = None
        
        # No stacktrace found for following nodes
        sym_constrain_range_30 = torch.ops.aten.sym_constrain_range.default(_local_scalar_dense_38, min = 0, max = 9223372036854775806)
        mul_36: "Sym(-u42)" = -1 * _local_scalar_dense_38
        le_28: "Sym(-u42 <= 0)" = mul_36 <= 0
        _assert_scalar_30 = torch.ops.aten._assert_scalar.default(le_28, 'Deferred runtime assertion failed -u42 <= 0')
```

It is clear that `_local_scalar_dense_38` is u42 here, but it has instead been assigned u356 above. This renumbering typically occurs when we retrace the graph. We observe the creation of u356 but it is never refined:

```
[rank0]:I0411 06:40:55.137000 140470255747520 torch/fx/experimental/symbolic_shapes.py:2580] [0/0_1] create_unbacked_symint u356 [-9223372036854775808, 9223372036854775807] (_subclasses/fake_impls.py:288 in local_scalar_dense)
[rank0]:I0411 06:40:55.137000 140470255747520 torch/fx/experimental/symbolic_shapes.py:2580] [0/0_1] For C++ stack trace, run with TORCHDYNAMO_EXTENDED_DEBUG_CPP=1
[rank0]:I0411 06:40:55.140000 140470255747520 torch/fx/experimental/symbolic_shapes.py:3884] [0/0_1] set_replacement u336 = u42 (find) ValueRanges(lower=0, upper=9223372036854775806, is_bool=False)
[rank0]:I0411 06:40:55.149000 140470255747520 torch/fx/experimental/symbolic_shapes.py:456] [0/0_1] _constrain_symbol_range u356 [0, 9223372036854775806]
```

This is the kind of problem that https://github.com/pytorch/pytorch/pull/120816 is supposed to help with. I am going to retry the flow with this PR.

Reading other occurences of u42, I observe another interesting phenomenon. In the AOTAutograd forward graph, I see:

```
        # File: /data/users/ezyang/fbsource/buck-out/v2/gen/fbcode/410ce23094e3dfef/torchrec/distributed/tests/__test_pt2_train__/test_pt2_train#link-tree/torchrec/sparse/jagged_tensor.py:239 in _permute_tensor_by_segments, code: output = torch.ops.fbgemm.keyed_jagged_index_select_dim1(
        keyed_jagged_index_select_dim1_2 = torch.ops.fbgemm.keyed_jagged_index_select_dim1.default(wait_tensor_1, wait_tensor_3, asynchronous_complete_cumsum_4, convert_element_type_5, 4);  asynchronous_complete_cumsum_4 = None
        getitem_8: "i64[u42 + u43 + u44 + u45]" = keyed_jagged_index_select_dim1_2[0];  keyed_jagged_index_select_dim1_2 = None
        
        # File: /data/users/ezyang/fbsource/buck-out/v2/gen/fbcode/410ce23094e3dfef/torchrec/distributed/tests/__test_pt2_train__/test_pt2_train#link-tree/torchrec/sparse/jagged_tensor.py:82 in _to_offsets, code: return torch.ops.fbgemm.asynchronous_complete_cumsum(lengths)
        asynchronous_complete_cumsum_7: "i64[5]" = torch.ops.fbgemm.asynchronous_complete_cumsum.default(sub)
        
        # File: /data/users/ezyang/fbsource/buck-out/v2/gen/fbcode/410ce23094e3dfef/torchrec/distributed/tests/__test_pt2_train__/test_pt2_train#link-tree/torchrec/sparse/jagged_tensor.py:239 in _permute_tensor_by_segments, code: output = torch.ops.fbgemm.keyed_jagged_index_select_dim1(
        keyed_jagged_index_select_dim1_3 = torch.ops.fbgemm.keyed_jagged_index_select_dim1.default(wait_tensor_2, sub, asynchronous_complete_cumsum_7, convert_element_type_5, 4)
        getitem_10: "i64[u46 + u47 + u48 + u49]" = keyed_jagged_index_select_dim1_3[0];  keyed_jagged_index_select_dim1_3 = None
        
        # File: /data/users/ezyang/fbsource/buck-out/v2/gen/fbcode/410ce23094e3dfef/torchrec/distributed/tests/__test_pt2_train__/test_pt2_train#link-tree/torchrec/sparse/jagged_tensor.py:2790 in _dist_init_per_key_tensor_compute, code: x = stride_per_key_per_rank[i][j].item()
        select_36: "i64[2]" = torch.ops.aten.select.int(device_put_6, 0, 0)
        select_37: "i64[]" = torch.ops.aten.select.int(select_36, 0, 0)
        _local_scalar_dense_38: "Sym(u42)" = torch.ops.aten._local_scalar_dense.default(select_37);  select_37 = None
```

The first line is in fact the first occurrence of u42. This is due to the following replacement:

```
[rank0]:I0411 06:40:21.768000 140470255747520 torch/fx/experimental/symbolic_shapes.py:3884] [0/0] set_replacement u40 = u42 + u43 + u44 + u45 (trivial_rhs) ValueRanges(lower=0, upper=36893488147419103224, is_bool=False)
```

However, this replacement is illegal: if we replace in this way, we now reference u42 before it is actually bound.

https://github.com/pytorch/pytorch/pull/120816 also potentially helps here, but I am not sure we actually know how to enforce this invariant.

### Versions

main

**Comment by ezyang** [comment]
I reran the internal repro with #120816 patched in and it still fails, which is promising because @ColinPeppler is also reporting failures post this patch:

```
  File "/data/users/ezyang/fbsource/buck-out/v2/gen/fbcode/410ce23094e3dfef/torchrec/distributed/tests/__test_pt2_train__/test_pt2_train#link-tree/torch/_
inductor/scheduler.py", line 1302, in __init__                                                                                                                self.compute_dependencies()                                                                                                                             File "/data/users/ezyang/fbsource/buck-out/v2/gen/fbcode/410ce23094e3dfef/torchrec/distributed/tests/__test_pt2_train__/test_pt2_train#link-tree/torch/_
inductor/scheduler.py", line 1542, in compute_dependencies                                                                                                
    s in unbacked_symbol_to_origin_node                                                                                                                   
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:                                                                                       
AssertionError: u312 not in {u352: ExternKernelSchedulerNode(name='buf7'
```

tlparse: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpYnd13H/rank_0/index.html

We observe a similar problem:

```
        # No stacktrace found for following nodes
        sym_size_int: "Sym(u352)" = torch.ops.aten.sym_size.int(getitem, 0)
        sym_constrain_range = torch.ops.aten.sym_constrain_range.default(sym_size_int, min = 0, max = 9223372036854775806)
        mul: "Sym(-u312)" = -1 * sym_size_int
        le: "Sym(-u312 <= 0)" = mul <= 0
        _assert_scalar = torch.ops.aten._assert_scalar.default(le, 'Deferred runtime assertion failed -u1 <= 0');  le = None
```

**Comment by ezyang** [comment]
Smoking gun:

```
[rank0]:I0411 14:51:09.524000 140344712118720 torch/fx/experimental/symbolic_shapes.py:2619] [0/0_1]   File "/data/users/ezyang/fbsource/buck-out/v2/gen/fbcode/410ce23094e3dfef/torchrec/distributed/tests/__test_pt2_train__/test_pt2_train#link-tree/fbgemm_gpu/sparse_ops.py", line 547, in keyed_jagged_index_select_dim1_abstract
[rank0]:I0411 14:51:09.524000 140344712118720 torch/fx/experimental/symbolic_shapes.py:2619] [0/0_1]     torch.index_select(lengths, 0, length_indices).sum().item()
```

**Comment by ezyang** [comment]
The root cause appears to be torch/_inductor/fx_utils.py is not setting up renamings when it recomputes fake tensors for nodes that involve unbacked SymInts. This is the "nightmare" case because we do not actually necessarily know if the new fake tensor corresponds to the old one, because a nontrivial optimization pass has occurred.

It's not clear which post grad pass caused the incremental update to occur.

```
[rank0]:I0411 14:51:09.524000 140344712118720 torch/fx/experimental/symbolic_shapes.py:2619] [0/0_1]   File "/data/users/ezyang/fbsource/buck-out/v2/gen/fbcode/410ce23094e3dfef/torchrec/distributed/tests/__test_pt2_train__/test_pt2_train#link-tree/torch/_inductor/compile_fx.py", line 464, in compile_fx_inner
[rank0]:I0411 14:51:09.524000 140344712118720 torch/fx/experimental/symbolic_shapes.py:2619] [0/0_1]     compiled_graph = fx_codegen_and_compile(
[rank0]:I0411 14:51:09.524000 140344712118720 torch/fx/experimental/symbolic_shapes.py:2619] [0/0_1]   File "/data/users/ezyang/fbsource/buck-out/v2/gen/fbcode/410ce23094e3dfef/torchrec/distributed/tests/__test_pt2_train__/test_pt2_train#link-tree/torch/_inductor/compile_fx.py", line 666, in fx_codegen_and_compile
[rank0]:I0411 14:51:09.524000 140344712118720 torch/fx/experimental/symbolic_shapes.py:2619] [0/0_1]     _recursive_post_grad_passes(gm, is_inference=is_inference)
[rank0]:I0411 14:51:09.524000 140344712118720 torch/fx/experimental/symbolic_shapes.py:2619] [0/0_1]   File "/data/users/ezyang/fbsource/buck-out/v2/gen/fbcode/410ce23094e3dfef/torchrec/distributed/tests/__test_pt2_train__/test_pt2_train#link-tree/torch/_inductor/compile_fx.py", line 240, in _recursive_post_grad_passes
[rank0]:I0411 14:51:09.524000 140344712118720 torch/fx/experimental/symbolic_shapes.py:2619] [0/0_1]     post_grad_passes(gm, is_inference)
```
